src/Dhall/Import.hs

```diff
@@ -665,17 +622,17 @@ loadStaticWith from_import ctx n expr₀ = case expr₀ of
                     --
                     -- There is no need to check expressions that have been
                     -- cached, since they have already been checked
-                    expr''' <- case Dhall.TypeCheck.typeWith ctx expr'' of
+                    expr''' <- case Dhall.TypeCheck.typeWith _startingContext expr'' of
                         Left  err -> throwM (Imported imports' err)
-                        Right _   -> return (Dhall.Core.normalizeWith n expr'')
-                    zoom cache (State.put $! Map.insert here expr''' m)
+                        Right _   -> return (Dhall.Core.normalizeWith (getReifiedNormalizer _normalizer) expr'')
+                    zoom cache (State.put $! Map.insert here expr''' _cache)
                     return expr'''
 
     case hash (importHashed import_) of
         Nothing -> do
             return ()
         Just expectedHash -> do
-            let actualHash = hashExpression expr
+            let actualHash = hashExpression _protocolVersion expr
```

Comment:
So the hashes guarding all imports, both direct and transitive, are governed by this one protocol version. Doesn't this risk ossifying the protocol version across the ecosystem?

Reply:
@quasicomputational: How about the following set of changes:

* Change `decode` to not require a version number
    * Instead, it just infers the version from the binary representation, which already includes the version string
    * In retrospect, I'm not sure why I made it require a `ProtocolVersion` argument :stuck_out_tongue: 
* Change semantic integrity checks to allow the user to optionally specify the protocol version string
    * i.e. `sha256:${HASH}:${PROTOCOL_VERSION}`
    * If the protocol version is unspecified then the interpreter assumes the latest protocol version string
* Remove the `--protocol-version` flag

Do you think that would work?

Reply:
That sounds reasonable. The other thing it occurs to me to do would be to allow many hashes and then have the importer pick the one with a version that it knows about, but that's probably putting the cart before the horse and making things too complex.

Any solution here that keeps dependents and dependencies from having to upgrade their hashes in lockstep would work, and, so long as decoders understand a few previous versions of the protocol version, this should work fine.

Reply:
Actually, I just realized that adding support for `sha256:${HASH}:${PROTOCOL_VERSION}` would be implementing a new feature without standardizing it.  What I will do in the interim is still remove support for the `--protocol-version` flag as I mentioned before, but wait to add support for the `:${PROTOCOL_VERSION}` extension to the semantic integrity check until it has been standardized.

This implies that I will delay standardizing any change that changes the binary protocol until we have first standardized support for pinning the protocol via the extension to the semantic integrity check.

Reply:
Actually, thinking on this more, we probably want to keep the `--protocol-version` flag, for two reasons:

* We minimally require the flag for the `dhall freeze` and `dhall hash` subcommands since the protocol version to use when hashing the expression(s) cannot be inferred from the input

* Even if we do eventually support specifying the protocol version on a per-import basis, we still need to specify a default protocol version when the protocol version is not pinned for an import.  Rather than hard-coded the default to be the latest version I think it makes sense to give users the ability to override the default protocol version from the command line.

So in other words, we'll still allow users to specify the protocol version on a per-import basis once that is standardized, but even after that we probably want a way for users to specify the default protocol version from the command line when otherwise unspecified.